src/main/GCNodeAllocator.cpp

```diff
@@ -0,0 +1,511 @@
+  }
+  // Test the upper 2 bits of the remaining bits.
+  if (remaining & 0xC) {
+    log2 += 2;
+    remaining >>= 2;
+  }
+  // Adjust for the final bits.
+  if (remaining > 0) {
+    log2 += remaining - 1;
+  }
+#endif // __has_builtin(__builtin_clz)
+
+  // Now make sure we go one above the highest set bit if the input is larger
+  // than that power of two:
+  if (size > (1 << log2)) {
+    log2 += 1;
+  }
+  return log2;
+}
+
+/**
+ * Allocate the small object arena, null out freelist head pointers,
+ * and initialize the hashtable.
+ */
+void rho::GCNodeAllocator::initialize() {
+  AllocatorSuperblock::allocateArena();
+
+  for (int i = 0; i < s_num_small_pools + s_num_medium_pools; ++i) {
+    // Initialize available superblock pointers.
+    s_superblocks[i] = nullptr;
+  }
+
+  for (int i = 0; i < s_num_freelists; ++i) {
+    // Initialize regular freelists.
+    s_freelists[i] = nullptr;
+#ifdef HAVE_ADDRESS_SANITIZER
+    // Initialize quarantine freelists.
+    s_quarantine[i] = nullptr;
+#endif
+  }
+
+  // Use a 16 bit hash initially.
+  s_alloctable = new rho::AllocationTable(16);
+}
+
+void* rho::GCNodeAllocator::allocate(size_t bytes) {
+  void* result = nullptr;
+#ifdef HAVE_ADDRESS_SANITIZER
+  // Increase size to include redzones.
+  bytes += 2 * s_redzone_size;
+#endif
+  unsigned actual_bytes;  // The actual allocation size
+      // (may be more than requested).
+  if (bytes <= 256) {
+    int size_class = (bytes + 7) / 8; // Computes ceil(bytes / 8).
+    if (size_class < 4) {
+      // Ensure at least 32-byte blocks for the small-object arena. This is
+      // required both to fit a FreeListNode (20 bytes), and to reduce the
+      // number of bytes needed for the fixed size bitset (as part of the
+      // constant SUPERBLOCK_HEADER_SIZE).
+      size_class = 4;
+    }
+    actual_bytes = size_class * 8;
+    result = removeFromFreelist(size_class);
+    if (!result) {
+      result = AllocatorSuperblock::allocateBlock(actual_bytes);
+      // If allocating in the small object arena fails, we continue
+      // on to using the medium block allocator. The allocation
+      // size will be increased to the minimum allocation size 64 bytes.
+    }
+  }
+  if (!result) {
+    // Now we try to allocate in a large superblock or, if the size
+    // exceeds the superblock threshold, use a separate allocation.
+    // These allocations are rounded up to the next power of two size.
+    unsigned size_log2 = next_log2_32(bytes);
+    if (size_log2 < 6) {
+      // Ensure a minimum allocation size of 64 bytes.
+      size_log2 = 6;
+    }
+    actual_bytes = 1 << size_log2;
+    unsigned size_class = AllocatorSuperblock::sizeClassFromSizeLog2(size_log2);
+    result = removeFromFreelist(size_class);
+    if (!result) {
+      if (size_log2 < s_num_medium_pools) {
+        result = AllocatorSuperblock::allocateLarge(size_log2);
+      } else {
+        result = new double[(1L << size_log2) / sizeof(double)];
```

Comment:
IIUC, you're rounding up the size of each of the largest allocations to the next power of two.  What impact does that have?  Might it be more problematic on some OSes or configurations than others?